### ISSUE TITLE
bash _direnv_hook clobbers last command's exit status

### ISSUE THREAD

**Issue by sykesm** [direnv/direnv#155]
The exit status from `eval` is the exit status of the command that was evaluated and the current bash hook uses `eval` to process `direnv export bash` ahead of an existing `$PROMPT_COMMAND`.  This approach as the side effect of eating the exit status of the last shell command that was executed.

For example, executing /usr/bin/false returns 1:
```
$ /usr/bin/false; echo $?
1
```
and executing /usr/bin/true returns 0:
```
$ /usr/bin/true; echo $?
0
```

Now it's fairly common for a bash prompt to provide some indication of whether or not the previous command was successful or not. A really basic prompt would simply print out the exit status of the previous command by referencing `$?`. For example:
```
$ prompt_command() { export PS1="[rc=$?]\\$ "; }
$ export PROMPT_COMMAND=prompt_command
[rc=0]$ 
```

With this prompt command in place, calling `true` or `false` results in the expected prompt:
```
[rc=0]$ true
[rc=0]$ false
[rc=1]$ 
```

If I then evaluate the direnv hook, the behavior changes:
```
[rc=1]$ eval "$(direnv hook bash)"
[rc=0]$ true
[rc=0]$ false
[rc=0]$ echo $?
1
```

You'll notice that the prompt's view of the exit status does not reflect the status of the previously executed command at the prompt.  This is due to the fact that `$PROMPT_COMMAND` now evaluates `direnv export bash` prior to `prompt_command`:
```
[rc=0]$ echo $PROMPT_COMMAND 
_direnv_hook;prompt_command
[rc=0]$ 
```

If I simply change the bash `_direnv_hook` function from:
```bash
_direnv_hook() {
  eval "$(direnv export bash)";
};
```
to
```bash
_direnv_hook() {
  local previous_exit_status=$?;
  eval "$(direnv export bash)";
  return $previous_exit_status;
}
```
things begin to behave as expected:
```
[rc=0]$ true
[rc=0]$ echo $?
0
[rc=0]$ false
[rc=1]$ echo $?
1
[rc=0]$ 
```

Basically, `_direnv_hook` should preserve the exit status at entry and restore it at exit. If it doesn't, it breaks a number of customized bash prompts.

**Comment by zimbatm** [comment]
This is how bugs should be reported, wonderful in depth analysis of the
problem, and a proposed solution.

What if direnv fails, wouldn't it make sense to override the exit status in
that case ?

**Comment by pwaller** [comment]
I don't understand, why doesn't this bug manifest for me?

**Comment by sykesm** [comment]
@zimbatm I don't know the failure modes of `direnv` so it's hard for me to say. If `direnv export` failed, I would probably still preserve the previous exit status but indicate the failure with messages to stderr.

@pwaller I'm not sure why it doesn't manifest for you or when it started manifesting for me. I only dug into it yesterday when I had a command fail and noticed I didn't have a visual indication of the problem in my prompt.

As a point of reference, I'm running Max OS X 10.10.1 with version `GNU bash, version 3.2.53(1)-release (x86_64-apple-darwin14)` but I see similar behavior under Linux with bash 4:

```
$ bash --version
GNU bash, version 4.2.24(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2011 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
vagrant@precise64:~$ prompt_command() { export PS1="[rc=$?]\\$ "; }
vagrant@precise64:~$ export PROMPT_COMMAND=prompt_command
[rc=0]$ true
[rc=0]$ false
[rc=1]$ _direnv_hook() { eval ""; };
[rc=0]$ export PROMPT_COMMAND=_direnv_hook;prompt_command
[rc=0]$ true
[rc=0]$ false
[rc=0]$ 
```

The use of `eval ""` simulates the eval of `direnv export bash` when there are no environment changes.

**Comment by pwaller** [comment]
I wonder if there is a behavioural difference in bash `3.2.53` and `4.3.11` with respect to prompt command and exit status.

**Comment by zimbatm** [comment]
I was able to reproduce the issue on both `3.2.53` and `4.3.11`. If the ordering in PROMPT_COMMAND is changed the issue goes away. In the doc I'm insisting of installing direnv last in the `~/.bashrc` because of issues like that: https://github.com/zimbatm/direnv#bash . Still worth fixing though.

**Comment by miketheman** [comment]
This has been bugging me for over a year now - and I only found the source of it this morning.
I opened a PR with the suggested solution after building and testing it locally. It makes my bash prompt happy again.
